Marlin/Regulator.cpp

```diff
@@ -9,23 +9,136 @@
+static void pressure_protection(float pressure, float target_pressure);
+
+//===========================================================================
+//============================= Public Functions ============================
+//===========================================================================
+
+/**
+ * Set the output pressure (in psi) of the regulator.
+ * @desired_pressure  The pressure that the function is attempting to 
+ *                    reach.
+ */
+void Regulator__SetOutputPressure(float desired_pressure) {
+  uint16_t digital_val = 0;
+  current_target_pressure = desired_pressure;
+  // Increasing pressure
+  if (desired_pressure > pressureRegulator()) {
+    // Add half of hysteresis range to desired pressure
+    desired_pressure += (REG_HYSTERESIS / 2.0);
+  }
+  // Decreasing pressure
+  else if (desired_pressure < pressureRegulator()) {
+    // Subtract half of hysteresis range from desired pressure
+    desired_pressure -= (REG_HYSTERESIS / 2.0);
+  }
+  digital_val = (uint16_t)(BITS_PER_PSI * (desired_pressure + MCP_CONST));
+  // Write value to DAC
+  DAC_write(MCP4725_I2C_ADDRESS, digital_val);
+}
 
-	digital_val = (uint16_t)(BITS_PER_PSI * (desired_pressure + MCP_CONST));
-	// Write value to DAC
-	DAC_write(MCP4725_I2C_ADDRESS, digital_val);
+
+/**
+ * Updates the protection function, called regularly in the main loop.
+ */
+void Regulator__Update() {
+  pressure_protection(current_target_pressure, pressureRegulator());
+}
+
+//===========================================================================
+//============================ Private Functions ============================
+//===========================================================================
+
+ /** 
+  * Error handler when marlin detects a regulator leak 
+  */
+static void _regulator_leak_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOL(  
+      "Pressure Leak Error. Canceling print, check the pressure hose");
+  SERIAL_EOL;
+  SERIAL_ECHOLN("// action:cancel");
+  SERIAL_EOL;
+  Regulator__SetOutputPressure(0.0);
+}
+
+ /** 
+  * Error handler when marlin detects a regulator pressure that's too high
+  */
+static void _regulator_runaway_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOL("Runaway Pressure Error. Canceling print.");
+  SERIAL_EOL;
+  SERIAL_ECHOLN("// action:cancel");
+  SERIAL_EOL;
+  Regulator__SetOutputPressure(0.0);
+}
+
+ /** 
+  * Allows us to determine if the pressure is too high or low, indicating
+  * a serious issue. Checks the target temperature vs. the desired, and 
+  * throws an error if outside the bounds for too long.
+  * @target_pressure  The pressure you're attempting to reach
+  * @pressure         The actual current pressure 
+  */
+static void pressure_protection(float target_pressure, float pressure) {
+  static const int period_seconds = REGULATOR_PROTECTION_TIME_S;
```

Comment:
Na, I get where you're going with it. The code could just be in the update function. I like keeping it separate like this in case we're going to add anything else to the update, but it could be moved up